domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt

```diff
@@ -0,0 +1,461 @@
+import org.oppia.android.util.data.DataProvider
+import org.oppia.android.util.data.DataProviders.Companion.toLiveData
+import org.oppia.android.util.data.DataProvidersInjector
+import org.oppia.android.util.data.DataProvidersInjectorProvider
+import org.oppia.android.util.logging.EnableConsoleLog
+import org.oppia.android.util.logging.EnableFileLog
+import org.oppia.android.util.logging.GlobalLogLevel
+import org.oppia.android.util.logging.LogLevel
+import org.robolectric.annotation.Config
+import org.robolectric.annotation.LooperMode
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * The base exploration id for every exploration used for testing [ExplorationCheckpointController].
+ * The exploration id of every test exploration will start with
+ * the string [BASE_TEST_EXPLORATION_ID].
+ */
+private const val BASE_TEST_EXPLORATION_ID = "test_exploration_"
+
+/**
+ * The base exploration title for every exploration used for testing
+ * [ExplorationCheckpointController]. The exploration title of every test exploration will start
+ * with the string [BASE_TEST_EXPLORATION_TITLE].
+ */
+private const val BASE_TEST_EXPLORATION_TITLE = "Test Exploration "
+
+/**
+ * Tests for [ExplorationCheckpointController].
+ *
+ * For testing this controller, checkpoints of hypothetical explorations are saved, updated,
+ * retrieved and deleted. These hypothetical explorations are referred to as "test explorations".
+ */
+@RunWith(AndroidJUnit4::class)
+@LooperMode(LooperMode.Mode.PAUSED)
+@Config(application = ExplorationCheckpointControllerTest.TestApplication::class)
+class ExplorationCheckpointControllerTest {
+
+  @Rule
+  @JvmField
+  val mockitoRule: MockitoRule = MockitoJUnit.rule()
+
+  @Inject
+  lateinit var testCoroutineDispatchers: TestCoroutineDispatchers
+
+  @Inject
+  lateinit var context: Context
+
+  @Inject
+  lateinit var explorationCheckpointController: ExplorationCheckpointController
+
+  @Mock
+  lateinit var mockResultObserver: Observer<AsyncResult<Any?>>
+
+  @Captor
+  lateinit var resultCaptor: ArgumentCaptor<AsyncResult<Any?>>
+
+  @Mock
+  lateinit var mockExplorationCheckpointObserver: Observer<AsyncResult<ExplorationCheckpoint>>
+
+  @Captor
+  lateinit var explorationCheckpointCaptor: ArgumentCaptor<AsyncResult<ExplorationCheckpoint>>
+
+  @Mock
+  lateinit var mockCheckpointDetailsObserver: Observer<AsyncResult<ExplorationCheckpointDetails>>
+
+  @Captor
+  lateinit var checkpointDetailsCaptor: ArgumentCaptor<AsyncResult<ExplorationCheckpointDetails>>
+
+  private val firstTestProfile = ProfileId.newBuilder().setInternalId(0).build()
+  private val secondTestProfile = ProfileId.newBuilder().setInternalId(1).build()
+
+  @Before
+  fun setUp() {
+    setUpTestApplicationComponent()
+  }
+
+  @Test
+  fun testController_saveCheckpoint_databaseNotFull_isSuccessfulWithDatabaseInCorrectState() {
+
+    val saveCheckpointLiveData = saveCheckpoint(firstTestProfile, 0).toLiveData()
```

Comment:
Nit here and elsewhere:
`val saveCheckpointLiveData = saveCheckpoint(firstTestProfile, index = 0).toLiveData()`

Reply:
done